crates/pico_macros/src/db.rs

```diff
@@ -0,0 +1,71 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput};
+
+pub fn derive_db(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let struct_name = input.ident;
+
+    let output = quote! {
+        impl ::pico_core::database::Database for #struct_name {
+            fn storage(&self) -> &impl ::pico_core::storage::Storage<Self> {
+                &self.storage
+            }
+
+            fn storage_mut(&mut self) -> &mut impl ::pico_core::storage::StorageMut<Self> {
+                &mut self.storage
+            }
+
+            fn current_epoch(&self) -> ::pico_core::epoch::Epoch {
+                use ::pico_core::storage::Storage;
+                self.storage().current_epoch()
+            }
+
+            fn increment_epoch(&mut self) -> ::pico_core::epoch::Epoch {
+                use ::pico_core::storage::StorageMut;
+                self.storage_mut().increment_epoch()
+            }
+
+            fn get<T: Clone + 'static>(&mut self, id: ::pico_core::source::SourceId<T>) -> T {
+                use ::pico_core::{storage::Storage, container::Container};
+                let time_calculated = self.storage()
+                    .sources()
+                    .get(&id.key)
+                    .expect("node should exist. This is indicative of a bug in Pico.")
+                    .time_calculated;
+                let current_epoch = self.current_epoch();
+                ::pico::memo::register_dependency(self, ::pico_core::node::NodeKind::Source(id.key), time_calculated, current_epoch);
+                self.storage()
+                    .source_values()
+                    .get(&id.key)
+                    .expect("value should exist. This is indicative of a bug in Pico.")
+                    .as_any()
+                    .downcast_ref::<T>()
+                    .expect("unexpected struct type. This is indicative of a bug in Pico.")
+                    .clone()
+            }
+
+            fn set<T>(&mut self, source: T) -> ::pico_core::source::SourceId<T>
+            where T: ::pico_core::source::Source + ::pico_core::dyn_eq::DynEq
+            {
+                use ::pico_core::{storage::StorageMut, container::Container};
+                let current_epoch = self.increment_epoch();
```

Comment:
I don't think we need to increment the epoch unless we modify an existing source.

That will mean that we can be more efficient in the following situation:

- call memoized function foo
- insert a new source
- call foo again

Now we can avoid iterating over foo's dependencies, since it was verified at the current epoch.

Thoughts? This could easily be a follow up as well